pkg/controllers/user/multiclusterapp/controller_multiclusterapp.go

```diff
@@ -420,52 +425,92 @@ func (m *MCAppController) deleteApps(mcAppName string, mcapp *v3.MultiClusterApp
-			}
-			if err := m.apps.DeleteNamespaced(app.Namespace, app.Name, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
-				return err
-			}
-			if err := m.namespaces.Delete(appWorkloadNamespace, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
-				return err
-			}
-			return nil
-		})
-	}
-
-	if err := g.Wait(); err != nil {
+	if err := m.delete(appsToDelete); err != nil {
 		return nil, err
 	}
 	return nil, nil
 }
 
-func (m *MCAppController) getAllAppsToDelete(mcAppName string) ([]*pv3.App, *v3.MultiClusterApp, error) {
+func (m *MCAppController) getAllApps(mcAppName string) ([]*pv3.App, error) {
 	// to get all apps, get all clusters first, then get all apps in all projects of all clusters
 	appsToDelete := []*pv3.App{}
 	set := labels.Set(map[string]string{multiClusterAppIDSelector: mcAppName})
 	clusters, err := m.clusterLister.List("", labels.NewSelector())
 	if err != nil {
-		return appsToDelete, nil, err
+		return appsToDelete, err
 	}
 	for _, c := range clusters {
 		projects, err := m.projectLister.List(c.Name, labels.NewSelector())
 		if err != nil {
-			return appsToDelete, nil, err
+			return appsToDelete, err
 		}
 		for _, p := range projects {
 			apps, err := m.appLister.List(p.Name, set.AsSelector())
 			if err != nil {
-				return appsToDelete, nil, err
+				return appsToDelete, err
 			}
 			appsToDelete = append(appsToDelete, apps...)
 		}
 	}
-	return appsToDelete, nil, err
+	return appsToDelete, err
+}
+
+func (m *MCAppController) reconcileTargetsForDelete(mcapp *v3.MultiClusterApp) error {
+	existingApps := map[string][]string{}
+	for ind, t := range mcapp.Spec.Targets {
+		if t.AppName == "" {
+			continue
+		}
+		split := strings.SplitN(t.ProjectName, ":", 2)
+		if len(split) != 2 {
+			return fmt.Errorf("invalid project name %s", t.ProjectName)
+		}
+		existingApps[t.AppName] = []string{split[1], strconv.Itoa(ind)}
+	}
+	allApps, err := m.getAllApps(mcapp.Name)
+	if err != nil {
+		return err
+	}
+	toDelete := []*pv3.App{}
+	for _, app := range allApps {
+		val, ok := existingApps[app.Name]
+		if !ok {
+			toDelete = append(toDelete, app)
+		} else if val[0] != app.Namespace {
+			toDelete = append(toDelete, app)
+			ind, err := strconv.Atoi(val[1])
+			if err != nil {
+				return err
+			}
+			mcapp.Spec.Targets[ind].AppName = ""
+		}
+	}
+	return m.delete(toDelete)
+}
+
+func (m *MCAppController) delete(appsToDelete []*pv3.App) error {
+	var g errgroup.Group
+	for ind := range appsToDelete {
+		app := appsToDelete[ind]
+		g.Go(func() error {
+			var appWorkloadNamespace string
+			if app != nil {
+				appWorkloadNamespace = app.Spec.TargetNamespace
+			}
+			if err := m.apps.DeleteNamespaced(app.Namespace, app.Name, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err := m.namespaces.Delete(appWorkloadNamespace, &metav1.DeleteOptions{}); err != nil && !apierrors.IsNotFound(err) {
```

Comment:
do we attempt the namespace removal multiple times? in every go routine

Reply:
nvm, it's one ns per app

Reply:
for each app